npm.d.ts

```diff
@@ -0,0 +1,439 @@
+	Run **before** the package is prepared and packed, **only** on `npm publish`.
+	*/
+	prepublishOnly?: string;
+
+	/**
+	Run **before** a tarball is packed (on `npm pack`, `npm publish`, and when installing git dependencies).
+	*/
+	prepack?: string;
+
+	/**
+	Run **after** the tarball has been generated and moved to its final destination.
+	*/
+	postpack?: string;
+
+	/**
+	Run **after** the package is published.
+	*/
+	publish?: string;
+
+	/**
+	Run **after** the package is published.
+	*/
+	postpublish?: string;
+
+	/**
+	Run **before** the package is installed.
+	*/
+	preinstall?: string;
+
+	/**
+	Run **after** the package is installed.
+	*/
+	install?: string;
+
+	/**
+	Run **after** the package is installed and after `install`.
+	*/
+	postinstall?: string;
+
+	/**
+	Run **before** the package is uninstalled and before `uninstall`.
+	*/
+	preuninstall?: string;
+
+	/**
+	Run **before** the package is uninstalled.
+	*/
+	uninstall?: string;
+
+	/**
+	Run **after** the package is uninstalled.
+	*/
+	postuninstall?: string;
+
+	/**
+	Run **before** bump the package version and before `version`.
+	*/
+	preversion?: string;
+
+	/**
+	Run **before** bump the package version.
+	*/
+	version?: string;
+
+	/**
+	Run **after** bump the package version.
+	*/
+	postversion?: string;
+
+	/**
+	Run with the `npm test` command, before `test`.
+	*/
+	pretest?: string;
+
+	/**
+	Run with the `npm test` command.
+	*/
+	test?: string;
+
+	/**
+	Run with the `npm test` command, after `test`.
+	*/
+	posttest?: string;
+
+	/**
+	Run with the `npm stop` command, before `stop`.
+	*/
+	prestop?: string;
+
+	/**
+	Run with the `npm stop` command.
+	*/
+	stop?: string;
+
+	/**
+	Run with the `npm stop` command, after `stop`.
+	*/
+	poststop?: string;
+
+	/**
+	Run with the `npm start` command, before `start`.
+	*/
+	prestart?: string;
+
+	/**
+	Run with the `npm start` command.
+	*/
+	start?: string;
+
+	/**
+	Run with the `npm start` command, after `start`.
+	*/
+	poststart?: string;
+
+	/**
+	Run with the `npm restart` command, before `restart`. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
+	*/
+	prerestart?: string;
+
+	/**
+	Run with the `npm restart` command. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
+	*/
+	restart?: string;
+
+	/**
+	Run with the `npm restart` command, after `restart`. Note: `npm restart` will run the `stop` and `start` scripts if no `restart` script is provided.
+	*/
+	postrestart?: string;
+
+	[scriptName: string]: string | undefined;
+}
+
+/**
+Array of package names that is bundled when the package is publihed.
+*/
+export type BundledDependency = string[];
+
+/**
+Dependencies of the package. The version range is a string which has one or more space-separated descriptors. Dependencies can also be identified with a tarball or Git URL.
+*/
+export interface Dependency {
+	[packageName: string]: string;
+}
+
+export type PackageJson = {
```

Comment:
Wouldn't it be better if all the other types/interfaces were nested under this one? So `Dependency` was referred to as `PackageJson.Dependency`. I think that would be a better namespace than putting everything under `Npm`.

Reply:
I did the refactoring locally. The problem is that VS Code is no longer able to import the types via code action. So it's kind of a degraded experience.

VS Code uses TypeScript server under the hood, like most other IDEs to provide TS support.